src/simulation.jl

```
@@ -4,8 +4,10 @@
 Simulate a temporal point process `pp` on interval `[tmin, tmax)` using Ogata's algorithm.
 """
 function simulate_ogata(
-    rng::AbstractRNG, pp::AbstractPointProcess{M}, tmin::T, tmax::T
-) where {M,T<:Real}
+    rng::AbstractRNG, pp::AbstractPointProcess, tmin::T, tmax::T
+) where {T<:Real}
+    h = History(; times=T[], marks=[], tmin=tmin, tmax=tmax)
+    M = typeof(rand(mark_distribution(pp, zero(T), h)))
```

I don't think this is going to work in general, because the first `h.marks` has element type `Any`. At the moment, the package only contains very simple test cases where `mark_distribution` actually doesn't depend on `h`, so there is no way to check.
An elegant way out of this conundrum might be to say that the argument `h` is optional in functions like `mark_distribution` or `ground_intensity`. In other words, the implementation must support no `h` being passed, meaning that we compute the relevant quantity at the very beginning of a new history. What do you think?

---

In my point of view, the type of the marks produced by `mark_distribution` is inherited from the point process model `pp` rather than from the type of `h.marks`. Furthermore, I would say that `mark_distribution` should throw an error if `h.marks` cannot be converted to `Vector{M}` where `M` is the type of the marks of `pp`.

---

The meaning of my remark is that in a reasonable implementation of `mark_distribution`, it would _expect_ to get a concretely-typed vector of marks inside the history `h`. If that is not satisfied (as in your proposal where `h.marks isa Vector{Any}`), then all bets are off. The code may error, or it may be type-unstable.
In retrospect, I think this issue with initializing Ogata was the reason why I decided to include the mark type in the point process. Which is not so far away from you saying 

> the type of the marks produced by mark_distribution is inherited from the point process model pp rather than from the type of h.marks.

---

I understand your point. And it's also related with your comment on issue #49 ,
> > It can be inferred from mark_dist, no?

>Not necessarily, sometimes the only way would be to get a sample. But maybe we don't need to infer it when we don't have a history.

Nevertheless, I will try a last comment ;-).
I think that `convert(Vector{T}, Any[])` works for any type `T` and so I would say that your concerns is settled by automatic conversion, no ?

---

But which `T` would you pick?

---

Oups... you are right

So I guess that the best way is indeed to keep the mark type in `pp` as you did. I will come back to this.